xayn-ai/src/ranker/context.rs

```diff
@@ -1,119 +1,231 @@
-pub(crate) struct Context {
-    /// Average positive distance.
-    pos_avg: f32,
-    /// Maximum negative distance.
-    neg_max: f32,
+use std::{
+    collections::HashMap,
+    time::{Duration, SystemTime},
+};
+
+use displaydoc::Display;
+use thiserror::Error;
+
+use crate::{
+    coi::{
+        compute_coi_decay_factor,
+        find_closest_coi,
+        point::{CoiPoint, UserInterests},
+        RelevanceMap,
+    },
+    embedding::utils::Embedding,
+    ranker::{document::Document, Configuration},
+    utils::system_time_now,
+    CoiId,
+    DocumentId,
+};
+
+#[derive(Error, Debug, Display)]
+#[allow(clippy::enum_variant_names)]
+pub(crate) enum Error {
+    /// Not enough positive cois (expected {expected:?}, has {has:?})
+    NotEnoughPositiveCois { expected: u32, has: u32 },
+    /// Not enough negative cois (expected {expected:?}, has {has:?})
+    NotEnoughNegativeCois { expected: u32, has: u32 },
+    /// Failed to find the closest cois
+    FailedToFindTheClosestCois,
 }
 
-impl Context {
-    pub(crate) fn from_cois(cois: &[(DocumentId, CoiComponent)]) -> Self {
-        let cois_len = cois.len() as f32;
-        let pos_avg = cois.iter().map(|(_, coi)| coi.pos_distance).sum::<f32>() / cois_len;
-        let neg_max = cois
-            .iter()
-            .map(|(_, coi)| coi.neg_distance)
-            .fold(f32::MIN, f32::max); // NOTE f32::max considers NaN as smallest value
+/// Helper struct for [`find_closest_cois`].
+struct ClosestCois {
+    /// The ID of the positive centre of interest
+    pos_id: CoiId,
+    /// Distance from the positive centre of interest
+    pos_distance: f32,
+    pos_last_view: SystemTime,
+
+    /// Distance from the negative centre of interest
+    neg_distance: f32,
+    neg_last_view: SystemTime,
+}
 
-        Self { pos_avg, neg_max }
+fn find_closest_cois(
+    embedding: &Embedding,
+    user_interests: &UserInterests,
+    neighbors: usize,
+) -> Option<ClosestCois> {
+    let (pos_coi, pos_distance) = find_closest_coi(&user_interests.positive, embedding, neighbors)?;
+    let (neg_coi, neg_distance) = find_closest_coi(&user_interests.negative, embedding, neighbors)?;
+
+    ClosestCois {
+        pos_id: pos_coi.id(),
+        pos_distance,
+        pos_last_view: pos_coi.stats.last_view,
+        neg_distance,
+        neg_last_view: neg_coi.last_view,
     }
+    .into()
+}
+
+fn compute_score_for_embedding(
+    embedding: &Embedding,
+    user_interests: &UserInterests,
+    relevances: &mut RelevanceMap,
+    neighbors: usize,
+    horizon: Duration,
+    now: SystemTime,
+) -> Result<f32, Error> {
+    let cois = find_closest_cois(embedding, user_interests, neighbors)
+        .ok_or(Error::FailedToFindTheClosestCois)?;
+
+    let pos_decay = compute_coi_decay_factor(horizon, now, cois.pos_last_view);
+    let neg_decay = compute_coi_decay_factor(horizon, now, cois.neg_last_view);
+
+    relevances.compute_relevances(&user_interests.positive, horizon);
```

Comment:
at this point we probably should make `now: SystemTime` also an argument for the `compute_relevances()` function such that all time-dependent computations use the same `now`.